### PR TITLE
Add REINFORCE implementation tutorial

### PR DESCRIPTION
**Pull request by siddarth-c**
# Description

Created a new tutorial depicting the new .step() function of gymnasium v26 using PyTorch. REINFORCE is employed to solve Mujoco's Reacher.

## Type of change

Please delete options that are not relevant.

- [ ] This change requires a documentation update

# Checklist:

- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `pre-commit run --all-files` (see `CONTRIBUTING.md` instructions to set it up)
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: docs/tutorials/reinforce_reacher_gym_v26.py [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+
+        avg_reward = int(np.average(reinforce_rewards_across_episodes[-50:]))
+
+        if e % 100 == 0:
+            print('Episode:', e, 'Average Reward:', avg_reward)
+        
+    reinforce_rewards_across_seeds.append(reinforce_rewards_across_episodes)
+    print('-' * 20)
+
+df1 = pd.DataFrame(reinforce_rewards_across_seeds).melt()
```
Comment (pseudo-rnd-thoughts) [inline docs/tutorials/reinforce_reacher_gym_v26.py:234]: Can this be replaced by the `RecordEpisodeStatistics` wrappers and matplotlib

File: docs/tutorials/reinforce_reacher_gym_v26.py [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+warnings.filterwarnings("ignore")
+
+env = gym.make("Reacher-v4")
+
+class Policy_Network(nn.Module):
+
+    def __init__(self, observation_space, action_space) -> None:
+        super(Policy_Network, self).__init__()
+
+        """
```
Comment (pseudo-rnd-thoughts) [inline docs/tutorials/reinforce_reacher_gym_v26.py:54]: We use google style docstrings, could you update

File: docs/tutorials/reinforce_reacher_gym_v26.py [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+import pandas as pd
+import seaborn as sns
+import random
+
+import matplotlib.pyplot as plt
+from matplotlib.lines import Line2D
+plt.rcParams["figure.figsize"] = (10,5)
+
+import warnings
+warnings.filterwarnings("ignore")
```
Comment (pseudo-rnd-thoughts) [inline docs/tutorials/reinforce_reacher_gym_v26.py:45]: Why have you added this?

File: docs/tutorials/reinforce_reacher_gym_v26.py [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+
+env = gym.make("Reacher-v4")
+
+class Policy_Network(nn.Module):
```
Comment (pseudo-rnd-thoughts) [inline docs/tutorials/reinforce_reacher_gym_v26.py:49]: Could you add a description to explain what the policy network does, why this architecture, what its input and output is

File: docs/tutorials/reinforce_reacher_gym_v26.py [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+        shared_features = self.shared_net(x)
+
+        action_means = self.policy_mean_net(shared_features)
+        
+        action_stddevs = self.policy_stddev_net(shared_features)
+        action_stddevs = torch.log(1 + torch.exp(action_stddevs)) # Kind a activation function log(1 + exp(x))
+
+        return action_means, action_stddevs
+
+class REINFORCE():
```
Comment (pseudo-rnd-thoughts) [inline docs/tutorials/reinforce_reacher_gym_v26.py:97]: Could you add a description of how the `REINFORCE` algorithm works, link to the original paper, explain the purpose of the hyperparameters

File: docs/tutorials/reinforce_reacher_gym_v26.py [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+        loss.backward()
+        self.optimizer.step()
+
+
+        # Empty / zero out all episode-centric variables
+        self.probs = []
+        self.rewards = []
+        self.entropy = 0
+
+episodes = int(5e3)
```
Comment (pseudo-rnd-thoughts) [inline docs/tutorials/reinforce_reacher_gym_v26.py:180]: Could you explain how the training works?

**Review by pseudo-rnd-thoughts — CHANGES_REQUESTED** [review]
Thanks for the tutorial, it looks very helpful.
Could you fix the pre-commit issues and address the comments

**Comment by siddarth-c** [comment]
I have updated the files as per the readme.md
Will do better the next time, thanks for bearing with me!

**Comment by pseudo-rnd-thoughts** [comment]
No worries, currently when I build the tutorial, on the left hand side, all of the titles appear here. 
I believe it is the way that you have structured the sections that causes this. could you fix
In addition, the same issues that get reported when building still appear and the final figure doesn't appear

**Comment by siddarth-c** [comment]
I have modified the structure of the titles and code following the Blackjack tutorial. And running through the tests mentioned in readme.md I do not get any warnings.

**Comment by pseudo-rnd-thoughts** [comment]
@siddarth-c I have made a number of upgrades to the tutorials. There are only a couple more thing before we can merge
1. Replace the top image of the agent with a gif of the trained agent completing the environment
2. Could you add an image for the policy network should the data flow. i.e., input data -> shared network -> split to mean network and std network -> output

**Comment by siddarth-c** [comment]
Done with the mentioned changes

**Review by pseudo-rnd-thoughts — APPROVED** [review]
That policy network figure is very nice, to confirm, this is not copied from someone else and is your own creation.
Also for the top gif, is this from the final agent? The agent doesn't seem to do very well in the environment

**Comment by siddarth-c** [comment]
The policy learned via REINFORCE is not optimal in Reacher (despite extensive hyperparameter searches).
So I have changed the environment to the Inverted Pendulum, where it is able to learn the optimal policy (achieves max reward of 1000). I hope this change is acceptable.

And yes, the policy network was designed by me

**Review by pseudo-rnd-thoughts — APPROVED** [review]
Amazing, thank you for the tutorial, we would be interested in anymore tutorials that you create
